core/src/main/scala/org/apache/spark/ui/WebUI.scala

```diff
@@ -126,7 +126,11 @@ private[spark] abstract class WebUI(
   def bind(): Unit = {
     assert(serverInfo.isEmpty, s"Attempted to bind $className more than once!")
     try {
-      val host = Option(conf.getenv("SPARK_LOCAL_IP")).getOrElse("0.0.0.0")
+      val host = if (Utils.isClusterMode(conf)) {
```

Comment:
This is the same issue as with the RPC address. the preference should be given to the setting by YARN admins. Opening a port on an expected network is an additional vulnerability. That said, YARN's default is also 0.0.0.0 which users will get with this patch as well.

Reply:
Sorry, but this is still bothering me. It probably works fine in YARN mode because the UI is proxied by the RM. But that's not true for any other mode, and the default before was to bind to `0.0.0.0` and now it is binding to a specific hostname. And that may not be the public one in a multi-homed machine, so now the UI wouldn't be accessible without extra configuration.

Reply:
I can make the change YARN-only. 
However, I'd like to point out I don't see any documentation stating `0.0.0.0` as the default.

Reply:
Even if it's not explicitly documented, that's what the previous code did.

Reply:
Also, same question as before: why not just leave the code as is? It will still work on YARN.

Reply:
If we don't bind all listener sockets consistently I don't see a point in this PR (modulo bug fixes). This is supposed to simplify deployment in restricted environments and make it consistent with the documentation. If we leave one of this untreated we force users to keep `SPARK_LOCAL_IP`  in sync with related YARN settings.

Reply:
I thought the point of this PR is for the driver RPC endpoint to bind the correct address (since it doesn't support binding to 0.0.0.0)?

I don't see how the http server changes that.

Reply:
I formulated the problem more broadly in the title of the PR: "NM host for driver end points". It's not a intuitive default behavior to bind to `0.0.0.0` if the backend (YARN) is explicitly configured not to, and we need a mechanism to allow Spark to inherit the YARN-determined bind address on NM.
You convinced me that the client mode is less critical, and it's easy to override via environment of spark submit (after the bug fix). Although I'd prefer using bindAddress everywhere for consistency and as it's documented.

Reply:
There's a big different between the RPC endpoint and the web UI.

For the RPC endpoint, your change is fixing a potential issue. For the web UI, this change is potentially introducing a new issue in non-YARN mode. So please, revert this one change and all the rest is good to go.

Otherwise I can't merge this.

Reply:
How about the compromise I was suggesting about making the webUI bind change only for YARN. YARN's default bind address at the NM is still `0.0.0.0` but if we change it at the YARN level we get it for both Spark rpc and webUI as well.

Reply:
I'd rather not change it unless it's fixing a problem. I don't see a problem being fixed here. Also, we should avoid adding more and more cluster-specific logic.

Reply:
you should not force users to bind to unintended addresses without providing reasonable means to change that.

Reply:
You can set `SPARK_LOCAL_IP`.

If you really want to change this, you must not change the current default behavior, which is to bind to all interfaces. If you manage to do that without doing cluster-specific checks, then I'm ok with it. But none of your current solutions really pass that bar.

Reply:
at best, the way you would do this is like this on each worker:
```
<property>
<name>yarn.nodemanager.admin-env</name>
<value>MALLOC_ARENA_MAX=$MALLOC_ARENA_MAX,SPARK_LOCAL_IP=$NM_HOST</value>
</property>
```
which builds on the fact that NM_HOST is defined earlier on the launch script or 
or some other value or maybe 
```
<property>
<name>yarn.nodemanager.admin-env</name>
<value>MALLOC_ARENA_MAX=$MALLOC_ARENA_MAX,SPARK_LOCAL_IP=${yarn.nodemanager.bind-host}</value>
</property>
```
As previously mentioned if I have to break abstractions and intermix YARN worker settings with Spark environment (which will also unnecessarily passed to other apps)  the only thing we need from this patch is the fix of setting the hostname  in JettyUtils.scala

Reply:
Or, again, you can do nothing, leaving the existing code here, and nothing will be broken.

Reply:
I am sorry, it does not make sense to me to treat RPC and http endpoints inconsistently. Therefore I am removing the logic borrowed from YARN for RPC. Now we have a simpler PR. We can achieve what I need either with `appMasterEnv.SPARK_LOCAL_IP` or cluster-side config. At  the same time we keep the prior behavior as you requested.